lib/puppet/functions/deprecation.rb

```diff
@@ -0,0 +1,28 @@
+# Function to print deprecation warnings, it will display a provided message as well as the calling function.
+# The deprecation message will only be displayed once per caller.
+# It is affected by the puppet setting 'strict', which can be set to :error, :off and :warning (default)
+#
+require "puppet_x/puppetlabs/stdlib/deprecation_state"
+
+Puppet::Functions.create_function(:deprecation) do
+  dispatch :deprecation do
+    param 'String', :message
+  end
+
+  def deprecation(message)
+    called_by = Kernel.caller.first
+    # we only want to display a deprecation message once per caller.
+    unless PuppetX::Puppetlabs::Stdlib::DeprecationState.deprecation_warning_said(called_by)
+      # depending on configuration setting of strict
+      case Puppet.settings[:strict]
+      when :off
+        # do nothing 
+      when :error
+        Puppet.err("#{message}, called by #{called_by}")
+      else
+        Puppet.warning("#{message}, called by #{called_by}")
+      end
+      PuppetX::Puppetlabs::Stdlib::DeprecationState.add_deprecation_warning_said(called_by)
```

Comment:
Puppet already has call-once deprecation feature. No need to write another. Also, the deprecation per call site may not be the best - there is an api where a key can be provided that serves as a uniqueness key for the deprecation being issued.
A call may look like this:
```
            Puppet.warn_once(:unresolved_module_dependencies,
                             "unresolved_dependencies_for_module_#{from_module_data.name}",
                             msg)
```
A key, that can be used to denote this particular type of deprecation (for instance also adding it to disabled_warnings). The second is the uniqueness key - in the example, the deprecation can appear once per module name. The `msg` is the message text including any positional information that is formatted by the user/caller of the method.